src/blockdev.rs

```diff
@@ -492,38 +492,65 @@ impl Drop for Mount {
 
 #[derive(Debug)]
 pub struct SavedPartitions {
-    sector_size: Option<u64>,
+    sector_size: u64,
     partitions: Vec<(u32, GPTPartitionEntry)>,
 }
 
 impl SavedPartitions {
     pub fn new(disk: &mut File, filters: &[PartitionFilter]) -> Result<Self> {
-        let mut result = Self {
-            sector_size: None,
-            partitions: Vec::new(),
+        // try reading GPT
+        let maybe_gpt = match GPT::find_from(disk) {
+            Ok(gpt) => Some(gpt),
+            // no GPT on this disk, so no partitions to save
+            Err(gptman::Error::InvalidSignature) => None,
+            Err(e) => return Err(e).chain_err(|| "reading partition table")?,
         };
 
-        // skip reading partition table if we don't have an accept filter
-        if filters.is_empty() {
-            return Ok(result);
-        }
+        // determine sector size
+        let sector_size = if disk
+            .metadata()
+            .chain_err(|| "getting disk metadata")?
+            .file_type()
+            .is_block_device()
+        {
+            // We have a real disk, so treat it as the source of truth.
+            get_sector_size(&disk)?.get() as u64
+        } else if let Some(ref gpt) = maybe_gpt {
+            // No real disk, but an existing GPT.  Trust it.
+            gpt.sector_size
+        } else {
+            // We're not reading from a real disk, and there's no existing
+            // GPT.  This should only happen in unit tests.  Pick a default.
```

Comment:
It sounds like this could be inject from the outside, then. The real path would probe-or-fail, while the test cases would directly inject the relevant fake value.

Reply:
Good suggestion, thank you!